modules/gmc-agent/src/main/java/org/gridgain/service/CacheService.java

```diff
@@ -0,0 +1,204 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.events.Event;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.events.DiscoveryCustomEvent;
+import org.apache.ignite.internal.managers.discovery.DiscoveryCustomMessage;
+import org.apache.ignite.internal.managers.eventstorage.GridEventStorageManager;
+import org.apache.ignite.internal.processors.cache.DynamicCacheDescriptor;
+import org.apache.ignite.internal.processors.cache.GridCacheProcessor;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.apache.ignite.internal.processors.cache.query.GridCacheSqlIndexMetadata;
+import org.apache.ignite.internal.processors.cache.query.GridCacheSqlMetadata;
+import org.apache.ignite.internal.processors.query.schema.message.SchemaFinishDiscoveryMessage;
+import org.apache.ignite.internal.util.typedef.F;
+import org.gridgain.agent.WebSocketManager;
+import org.gridgain.dto.cache.CacheInfo;
+import org.gridgain.dto.cache.CacheSqlIndexMetadata;
+import org.gridgain.dto.cache.CacheSqlMetadata;
+
+import static org.apache.ignite.events.EventType.EVT_CACHE_STARTED;
+import static org.apache.ignite.events.EventType.EVT_CACHE_STOPPED;
+import static org.apache.ignite.internal.events.DiscoveryCustomEvent.EVT_DISCOVERY_CUSTOM_EVT;
+import static org.gridgain.agent.StompDestinationsUtils.buildClusterCachesInfoDest;
+import static org.gridgain.agent.StompDestinationsUtils.buildClusterCachesSqlMetaDest;
+
+/**
+ * Cache service.
+ */
+public class CacheService implements AutoCloseable {
+    /** Cache events. */
+    private static final int[] EVTS_CACHE = new int[] {EVT_CACHE_STARTED, EVT_CACHE_STOPPED};
+
+    /** Context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Websocket manager. */
+    private final WebSocketManager mgr;
+
+    /**
+     * @param ctx Context.
+     * @param mgr Websocket manager.
+     */
+    public CacheService(GridKernalContext ctx, WebSocketManager mgr) {
+        this.ctx = ctx;
+        this.mgr = mgr;
+        log = ctx.log(CacheService.class);
+
+        GridEventStorageManager evtMgr = ctx.event();
+
+        // Listener for cache metadata change.
+        evtMgr.addLocalEventListener(this::onDiscoveryCustomEvent, EVT_DISCOVERY_CUSTOM_EVT);
+        evtMgr.addLocalEventListener(this::onCacheEvents, EVTS_CACHE);
+    }
+
+    /**
+     * Send initial state of caches information.
+     */
+    public void sendInitialState() {
+        sendCacheInfo();
+    }
+
+    /**
+     * @param evt Event.
+     */
+    private void onCacheEvents(Event evt) {
+        sendCacheInfo();
+    }
+
+    /**
+     * @param evt Event.
+     */
+    private void onDiscoveryCustomEvent(Event evt) {
```

Comment:
As I remember you should return true if you want to continue receive events to listener

Reply:
It's true only for org.apache.ignite.IgniteEvents#localListen :)